### DISCUSSION TITLE
How to add the value of variable in other nodes into the kernel in the current node

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#21584] [Q&A General] [answered]
Suppose that there are two one-dimensional curves which is divided into n elements. One is block1 and the other is block2. 
I want to add this equation into block1 and block2's kernel.
![43cd5a18df34b11ae717146e1e9a9d0](https://user-images.githubusercontent.com/109257685/178892059-6636c460-dde9-4b24-a79a-dc4b69d6c2ef.png)
_w_ is the main variable in this equation, _U_(coupled variable) is the average velocity of block1 and block2, In block1, _pn_(coupled variable) is the pressure of block2, and in block2, _pn_ is the pressure of block1.

I dont know whether this can be done within moose, can anyone tell me?

Best regards,
Ethan

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are the elements where this coupling exists adjacent?

Guillaume

**Reply by GiudGiud** [reply]
like the two 1D lines, how are they positioned?

**Reply by Ethan-xj** [reply]
The lines are parallel and the length the number of elements are all the same. And the equation need to be solved involves two lines' parameters(in detail: line1's first element and line2's first element). In geometry, these two elements are not adjacant. The main variable(_w)_ in this equation is like a average value in form, but it need to be obtained by solving PDE. I'm not sure if I should add this equation into some line's kernel to solve this PDE, or maybe there's any other way?

**Reply by GiudGiud** [reply]
We dont have "line" kernels, we have elemental and nodal kernels. 

If you have sums over multiple elements that are not adjacent, it will be a little tricky to get the contributions to the Jacobian right for Newton's method. You can compute the sums using the postprocessor system, which will get you the contribution to the residual. 
with this one for example
https://mooseframework.inl.gov/source/postprocessors/ElementIntegralVariablePostprocessor.html

Can the two lines just the same line or is there any reason they must be separate?
If they are the same lines with all the variables living on that same 1D mesh, it's much easier to write each variable into the other variable's equation.

**Reply by Ethan-xj** [reply]
Thank you for your relpy!
The two lines have the same variables(p, v and T) and in geometry and mesh they are same too. But the value of the variables should be different. I'm not sure if this could be called "they are the same lines with all the variables living on that same 1D mesh".
So if the two lines are the same how can I write each variable into the other variable's equation?

**Reply by GiudGiud** [reply]
You may use the Coupleable interface to write the local variable value of each variable into the equation.
https://mooseframework.inl.gov/moose/source/interfaces/Coupleable.html

For non local values it's more complicated. I would do a first try with an integral postprocessor. Then if that doesnt work we can explore constraints, mortar and other non-local methods

**Reply by Ethan-xj** [reply]
OK I will have a try. Thanks for your patient solution!
